torch/csrc/autograd/record_function.h

```diff
@@ -26,32 +26,35 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
-
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
+  RecordFunction() {}
+
+  explicit RecordFunction(Function* fn);
 
   explicit RecordFunction(
       std::string name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
 
   explicit RecordFunction(
       const char* name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+      int64_t current_sequence_nr = -1);
 
-  explicit RecordFunction(
+  void before(
+      Function* fn,
+      c10::optional<at::ArrayRef<c10::IValue>> inputs = c10::nullopt);
+
+  void before(
       std::string name,
-      GetPackedInputsCallback cb) : RecordFunction(name, -1, cb) {}
+      c10::optional<at::ArrayRef<c10::IValue>> inputs = c10::nullopt,
+      int64_t current_sequence_nr = -1);
 
-  explicit RecordFunction(
+  void before(
```

Comment:
Why do we need those functions by the way? They don't seem to simplify the code at all compared to what we had before (i.e. actual constructors).

Reply:
We use them because in some cases a switch in a macro would determine that we don't need to pass information (e.g. there're no callbacks used)